### PR TITLE
Fix possible runaway mutex in ManagedWebSocket.ReceiveAsyncPrivate()

### PR DESCRIPTION
**Pull request by yirkha**
I've been trying to remove some legacy code in our codebase that is doing a lot of extra work to avoid obscure deadlock issues observed with the WebSocket class many years ago, in .NET Framework and the native websocket.dll.

I wanted to prove to my colleagues that there were no such issues today anymore with the fully managed implementation in .NET 8, so I wrote a bunch of stress tests around ManagedWebSocket, doing things like Dispose() while ReceiveAsync()/SendAsync()/CloseAsync() are pending etc. Unfortunately, I have uncovered a new problem instead.

Quoting from the commit message:

> Commit 21f07e17 (Fixing concurrency issue in ManagedWebSocket ... #87966, 2023-07-18) has fixed some cases when calling Dispose() on a ManagedWebSocket could cause corruption of the (de)compressor state if there were some ongoing receive/send operations happening in parallel. It has also introduced some additional dispose checks to the class.
>
> One of those, added to ReceiveAsyncPrivate(), is buggy in the sense that it is done between acquiring _receiveMutex and entering the try/finally block that always releases it again on the way out. The _receiveMutex can thus potentially get stuck in the taken state and cause other tasks to never complete or deadlock.
>
> The solution is to perform the dispose check inside the try/finally block, to ensure the mutex is released correctly even in this case.

I have looked at other places where _receiveMutex/_sendMutex are manually handled like this and I have not seen a similar problem anywhere else.

----

Note there are no tests added.

I tried many different approaches, different order of the operations, more fine-tuning of WebSocketTestStream, but nope, it seems hard to reproduce this consistently. Even my original test that uncovered this failed on "run until failure" iteration 692/1000 only. And to debug what happened more reliably, I had to make a tight loop to do the same steps many times faster.

I think it could be reproduced reasonably well by calling ReceiveAsync() multiple time in parallel &ndash; or at least _twice_ on top of CloseAsync() &ndash; but it feels wrong to test any assertions about something that is explicitly called out in the API/docs as _not supported_ and _undefined behavior_.

Hence I rather present the fix raw like this. The change itself is trivial, and the maintenance cost of having an extra test for something as specific as this might be bigger than its usefulness.

But let me know if you would like a test nevertheless and have better ideas, I'll whip something up. Thanks

### CODE REVIEW

File: src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs [hunk 1]
```diff
@@ -705,10 +705,10 @@ private async ValueTask<TResult> ReceiveAsyncPrivate<TResult>(Memory<byte> paylo
             try
             {
                 await _receiveMutex.EnterAsync(cancellationToken).ConfigureAwait(false);
-                ObjectDisposedException.ThrowIf(_disposed, typeof(WebSocket));
-
                 try
                 {
+                    ObjectDisposedException.ThrowIf(_disposed, typeof(WebSocket));
```
Comment (filipnavara) [inline src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:710]: If I am reading this correctly, this exception will end up in the `catch` block and exception of a different type will be thrown instead.

Comment (filipnavara) [reply src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs:710]: Nevermind, misread the indentation on GitHub.

**Comment by CarnaViire** [comment]
cc @zlatanov @stephentoub fyi

**Review by zlatanov — APPROVED** [review]

**Comment by zlatanov** [comment]
Thank you for spotting this @yirkha. Here is a reproducible test that should verify the fix. Can you add it in the tests, please?

```c#
using var stream = new WebSocketTestStream();
var websocket = WebSocket.CreateFromStream(stream, new WebSocketCreationOptions());

Task r1 = websocket.ReceiveAsync(new Memory<byte>(new byte[1]), default).AsTask();
Task r2 = websocket.ReceiveAsync(new Memory<byte>(new byte[1]), default).AsTask();
Task r3 = websocket.ReceiveAsync(new Memory<byte>(new byte[1]), default).AsTask();

websocket.Dispose();

await Assert.ThrowsAsync<WebSocketException>(async () => await r1.WaitAsync(TimeSpan.FromSeconds(1)));
await Assert.ThrowsAsync<WebSocketException>(async () => await r2.WaitAsync(TimeSpan.FromSeconds(1)));
await Assert.ThrowsAsync<WebSocketException>(async () => await r3.WaitAsync(TimeSpan.FromSeconds(1)));
```

Without the fix, the 3rd assert will fail with TimeoutException. After the fix, the behavior of all receives should be the same.

**Comment by yirkha** [comment]
@zlatanov 
> Here is a reproducible test that should verify the fix. Can you add it in the tests, please?

Yeah, sure, that's an obvious way to validate it.

But as I said, it feels a bit weird to add a test like this if the [public documentation](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.websocket.receiveasync?view=net-8.0) clearly states
> Exactly (...) one receive is supported on each WebSocket object in parallel.
> Issuing multiple receives at the same time is _not supported_ and will result in an _undefined behavior_.

So I'd just like to double check if tying down the API behavior to the specifics of the current implementation like this is okay by the maintainers 🙂

**Comment by zlatanov** [comment]
You make a good point. I'd leave it out of the tests then, but would run the code I've given just to make sure the fix is correct and is working as expected even if it's obvious 😄

**Comment by yirkha** [comment]
Ack, works as expected
Screenshot from VS test run

**Comment by wfurt** [comment]
can you add the test to the PR @yirkha? (and sync up with main)

**Comment by yirkha** [comment]
Yes. I had to wait for some build break in main to be resolved, pushed now

**Review by CarnaViire — APPROVED** [review]
Thanks!

**Comment by CarnaViire** [comment]
Failing tests are unrelated known issues.

**Pull request merged**